packages/git/src/node/git-repository-watcher.ts

```diff
@@ -42,55 +45,109 @@ export class GitRepositoryWatcher implements Disposable {
-    protected schedule(initial: boolean = false, delay = initial ? 0 : 5000): void {
-        if (initial && !this.initial) {
-            this.initial = true;
-            this.clear();
+                if (this.gitIgnoreTester === undefined || !this.gitIgnoreTester.some(simplePattern => this.testAgainstSimplePattern(simplePattern, relativePath))) {
+                    this.lazyRefresh();
+                }
+            }
         }
-        if (this.watchTimer) {
-            return;
+    }
+
+    protected lazyRefresh: () => Promise<void> = debounce(() => this.syncStatus(false), 500);
+
+    private async parseGitIgnoreFile(repositoryUri: string) {
+        const response = await this.filesystem.resolveContent(repositoryUri + '/.gitignore', { encoding: 'utf8' });
+        const patterns = response.content.split(/\r\n|\r|\n/);
+        this.gitIgnoreTester = this.extractSafeIgnores(patterns);
+    }
+
+    /**
+     * Returns a list of patterns that are just a single folder or file name at the root.
+     *
+     * Testing against a .gitignore file is hard.  However we don't have to do a perfect job.
+     * If we just ignore files in the build directories then this will substantially improve
+     * performance.  It does not matter if we let some files through that should be ignored
+     * because the Git status call will ignore them.  We do have to be sure not to ignore
+     * anything that should not be ignored.
+     *
+     * @param patterns the raw patterns from the .gitignore file
+     * @return a list of simple root file names that can safely be ignored
+     */
+    private extractSafeIgnores(patterns: string[]): string[] {
+        // If any lines begin with !, just keep it simple and ignore nothing
+        if (patterns.some(pattern => pattern.startsWith('!'))) {
+            return [];
         }
-        this.watchTimer = setTimeout(async () => {
-            this.watchTimer = undefined;
-            const syncInitial = this.initial;
-            this.initial = false;
-            if (await this.syncStatus(syncInitial)) {
-                this.schedule();
-            }
-        }, delay);
+
+        return patterns.filter(pattern => pattern.match(/^[\w-.]+$/));
+    }
+
+    private testAgainstSimplePattern(simplePattern: string, relativePath: string) {
+        return relativePath === simplePattern
+            || relativePath.startsWith(simplePattern + '/')
+            || relativePath.startsWith(simplePattern + '\\');
     }
-    protected clear(): void {
-        if (this.watchTimer) {
-            clearTimeout(this.watchTimer);
-            this.watchTimer = undefined;
+
+    protected readonly toDispose = new DisposableCollection();
+    async watch(): Promise<void> {
+        const notCurrentlyWatching = this.toDispose.disposed;
+        if (notCurrentlyWatching) {
+            const repositoryUri = this.options.repository.localUri;
+
+            this.fileSystemWatcher.watchFileChanges(repositoryUri).then(watcher =>
```

Comment:
It doubles amount of linux file handlers we need to watch each repository, since they are already by the frontend.

How about watching watch only changes to `.gitignore` file? `nsfw` also is reliable at watching the single file.

Reply:
Also, it increases amount of processes. One more for each repo.